From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Theo Paris <theo@theoparis.com>
Date: Tue, 15 Sep 2026 20:21:36 -0700
Subject: [PATCH 02/11] arm64: add OpenDarwin board support (QEMU, SUPERBIRD,
 IPAD41)

Add ARM64 board targets that boot on non-Apple-silicon hardware:

  * Board headers (QEMU.h, SUPERBIRD.h, IPAD41.h), board_config.h
    wiring, and the pexpert/arm64 Makefile.
  * A minimal GICv2 MMIO driver (pexpert/gic) and the A7 AICv1
    skeleton (pexpert/aic) with a shared gic.h interface.
  * Board plumbing in the arm64 pmap, proc_reg, caches, machine
    routines, sleh, commpage, zalloc zone-map sizing, and vm_param
    addressing for 4K-page non-Apple boards.
  * Stub translation units for Apple-silicon-only hardware absent from
    this open-source drop (AMCC/CTRR, SART, UAT, tunables).
  * New MASTER.arm64.MacOSX + MASTER.arm64 OSS-hardware config and
    arm64 export-list updates.

Extracted from the OpenDarwin ARM64 platform-support work.
---
 bsd/dev/arm/km.c                              |   2 +-
 config/MASTER.arm64                           |   2 +-
 config/MASTER.arm64.MacOSX                    |   9 +-
 config/Private.arm64.exports                  |  65 --------
 iokit/Kernel/IOKitKernelInternal.h            |   4 +
 osfmk/arm/commpage/commpage_asm.s             |  18 +-
 osfmk/arm/pmap/pmap.c                         |  23 +++
 osfmk/arm/pmap/pmap.h                         |   2 +-
 osfmk/arm/pmap/pmap_data.h                    |  20 +++
 osfmk/arm/pmap/pmap_misc.c                    |   8 +
 osfmk/arm64/amcc_rorgn_common.c               |  14 ++
 osfmk/arm64/amcc_rorgn_ppl.c                  |  14 ++
 osfmk/arm64/amcc_rorgn_ppl_amcc.c             |  13 ++
 osfmk/arm64/amcc_rorgn_pv_ctrr.c              |  12 ++
 osfmk/arm64/arm_vm_init.c                     |   2 +
 osfmk/arm64/caches_asm.s                      |  21 +++
 .../corecrypto/arm64_isa_compatibility.h      |   2 +-
 osfmk/arm64/machine_routines.c                |  22 ++-
 osfmk/arm64/ppl/sart.c                        |  21 +++
 osfmk/arm64/ppl/sart.h                        |  19 +++
 osfmk/arm64/ppl/uat.h                         |  11 ++
 osfmk/arm64/proc_reg.h                        |  22 ++-
 osfmk/arm64/sleh.c                            |   4 -
 osfmk/arm64/tunables/tunables.s               |  13 ++
 osfmk/conf/files.arm64                        |   1 +
 osfmk/kern/zalloc.c                           |  13 +-
 osfmk/mach/arm/vm_param.h                     |  14 ++
 pexpert/aic/aic.c                             |  96 +++++++++++
 pexpert/arm/pe_init.c                         |  19 +--
 pexpert/arm/pe_kprintf.c                      |  17 +-
 pexpert/conf/files.arm64                      |   3 +
 pexpert/gic/gic.c                             | 156 ++++++++++++++++++
 pexpert/pexpert/arm64/IPAD41.h                |  80 +++++++++
 pexpert/pexpert/arm64/Makefile                |   6 +-
 pexpert/pexpert/arm64/QEMU.h                  | 110 ++++++++++++
 pexpert/pexpert/arm64/SUPERBIRD.h             |  95 +++++++++++
 pexpert/pexpert/arm64/VMAPPLE.h               |   2 +
 pexpert/pexpert/arm64/board_config.h          |  48 ++++++
 pexpert/pexpert/arm64/gic.h                   |  30 ++++
 39 files changed, 935 insertions(+), 98 deletions(-)
 create mode 100644 osfmk/arm64/amcc_rorgn_common.c
 create mode 100644 osfmk/arm64/amcc_rorgn_ppl.c
 create mode 100644 osfmk/arm64/amcc_rorgn_ppl_amcc.c
 create mode 100644 osfmk/arm64/amcc_rorgn_pv_ctrr.c
 create mode 100644 osfmk/arm64/ppl/sart.c
 create mode 100644 osfmk/arm64/ppl/sart.h
 create mode 100644 osfmk/arm64/ppl/uat.h
 create mode 100644 osfmk/arm64/tunables/tunables.s
 create mode 100644 pexpert/aic/aic.c
 create mode 100644 pexpert/gic/gic.c
 create mode 100644 pexpert/pexpert/arm64/IPAD41.h
 create mode 100644 pexpert/pexpert/arm64/QEMU.h
 create mode 100644 pexpert/pexpert/arm64/SUPERBIRD.h
 create mode 100644 pexpert/pexpert/arm64/gic.h

diff --git a/bsd/dev/arm/km.c b/bsd/dev/arm/km.c
index 3acfc7b5..60737d79 100644
--- a/bsd/dev/arm/km.c
+++ b/bsd/dev/arm/km.c
@@ -41,7 +41,7 @@ struct tty     *km_tty[1] = { 0 };
 /*
  * 'Global' variables, shared only by this file and kmDevice.m.
  */
-int             initialized = 0;
+static int      initialized = 0;
 
 static int      kmoutput(struct tty * tp);
 static void     kmstart(struct tty * tp);
diff --git a/config/MASTER.arm64 b/config/MASTER.arm64
index 55d51b55..533c01b6 100644
--- a/config/MASTER.arm64
+++ b/config/MASTER.arm64
@@ -24,7 +24,7 @@
 #  BSD_RELEASE =    [ BSD_BASE no_printf_str no_kprintf_str secure_kernel ]
 #  BSD_DEV =        [ BSD_BASE config_netboot config_imgsrc_access config_coredump config_ucoredump pgo config_vnguard ]
 #  BSD_DEBUG =      [ BSD_BASE config_netboot config_imgsrc_access config_coredump config_ucoredump pgo config_vnguard ]
-#  FILESYS_BASE =   [ devfs fifo fs_compression config_protect config_mnt_rootsnap config_triggers config_fse routefs namedstreams config_dataless_files bindfs]
+#  FILESYS_BASE =   [ devfs fifo fs_compression config_protect config_mnt_rootsnap config_triggers config_fse routefs namedstreams config_dataless_files bindfs mockfs]
 #  FILESYS_RELEASE= [ FILESYS_BASE ]
 #  FILESYS_DEV =    [ FILESYS_BASE config_union_mounts fdesc ]
 #  FILESYS_DEBUG =  [ FILESYS_BASE config_union_mounts fdesc ]
diff --git a/config/MASTER.arm64.MacOSX b/config/MASTER.arm64.MacOSX
index 69219cc4..17f9728e 100644
--- a/config/MASTER.arm64.MacOSX
+++ b/config/MASTER.arm64.MacOSX
@@ -14,7 +14,9 @@
 #  Standard Apple OS Configurations:
 #  -------- ----- -- ---------------
 #
-#if MASTER_CONFIG_ENABLE_EXCLAVES
+#if MASTER_CONFIG_OSS_HARDWARE
+#  ARM_EXTRAS_BASE = [ config_pmap_ppl nos_arm_asm nos_arm_pmap ]
+#elif MASTER_CONFIG_ENABLE_EXCLAVES
 #  ARM_EXTRAS_BASE = [ config_sptm config_sptm_iommu config_exclaves ]
 #elif MASTER_CONFIG_ENABLE_SPTM
 #  ARM_EXTRAS_BASE = [ config_sptm config_sptm_iommu ]
@@ -29,7 +31,7 @@
 #  BSD_RELEASE =    [ BSD_BASE ]
 #  BSD_DEV =        [ BSD_BASE config_vnguard ]
 #  BSD_DEBUG =      [ BSD_BASE config_vnguard ]
-#  FILESYS_BASE =   [ devfs fdesc fifo fs_compression config_protect config_mnt_rootsnap config_rosv_startup config_mount_vm config_mount_prebootrecovery config_basesystemroot config_fse quota namedstreams config_imgsrc_access config_triggers config_ext_resolver config_searchfs config_volfs config_appledouble nullfs config_mnt_suid config_firmlinks config_dataless_files config_union_mounts config_file_leases ]
+#  FILESYS_BASE =   [ devfs fdesc fifo fs_compression config_protect config_mnt_rootsnap config_rosv_startup config_mount_vm config_mount_prebootrecovery config_basesystemroot config_fse quota namedstreams config_imgsrc_access config_triggers config_ext_resolver config_searchfs config_volfs config_appledouble nullfs config_mnt_suid config_firmlinks config_dataless_files config_union_mounts config_file_leases mockfs ]
 #  FILESYS_RELEASE= [ FILESYS_BASE ]
 #  FILESYS_DEV =    [ FILESYS_BASE config_iocount_trace ]
 #  FILESYS_DEBUG =  [ FILESYS_BASE config_iocount_trace ]
@@ -58,7 +60,7 @@
 #  LIBKERN_RELEASE =[ LIBKERN_BASE zlib ]
 #  LIBKERN_DEV =    [ LIBKERN_BASE zlib iotracking ]
 #  LIBKERN_DEBUG =  [ LIBKERN_BASE zlib iotracking ]
-#if SOC_IS_VIRTUALIZED
+#if SOC_IS_VIRTUALIZED || MASTER_CONFIG_OSS_HARDWARE
 #  CPU_COUNTERS_BASE = [ ]
 #else /*!SOC_IS_VIRTUALIZED*/
 #  CPU_COUNTERS_BASE = [ config_cpu_counters config_pervasive_cpi config_pervasive_energy ]
@@ -102,6 +104,7 @@
 #  SECURITY_DEBUG =   [ SECURITY_BASE ]
 #  BASE =           [ PF VPN NFS ]
 #  RELEASE =        [ BASE KERNEL_RELEASE BSD_RELEASE FILESYS_RELEASE SKYWALK_RELEASE NETWORKING_RELEASE IOKIT_RELEASE LIBKERN_RELEASE    PERF_DBG_RELEASE MACH_RELEASE SCHED_RELEASE VM_RELEASE SECURITY_RELEASE ]
+#  RELEASE_ASSERT = [ RELEASE mach_assert ]		# like LLVM's Release+Asserts: optimized release build with live assertions
 #  DEVELOPMENT =    [ BASE KERNEL_DEV     BSD_DEV     FILESYS_DEV     SKYWALK_DEV     NETWORKING_DEV     IOKIT_DEV     LIBKERN_DEV        PERF_DBG_DEV     MACH_DEV     SCHED_DEV     VM_DEV     SECURITY_DEV     ]
 #  DEBUG =          [ BASE KERNEL_DEBUG   BSD_DEBUG   FILESYS_DEBUG   SKYWALK_DEBUG   NETWORKING_DEBUG   IOKIT_DEBUG   LIBKERN_DEBUG      PERF_DBG_DEBUG   MACH_DEBUG   SCHED_DEBUG   VM_DEBUG   SECURITY_DEBUG   ]
 #  KASAN =          [ BASE KERNEL_DEV     BSD_DEV     FILESYS_DEV     SKYWALK_DEV     NETWORKING_DEV     IOKIT_DEV     LIBKERN_DEV        PERF_DBG_KASAN   MACH_DEV     SCHED_DEV     VM_KASAN   SECURITY_DEV     ]
diff --git a/config/Private.arm64.exports b/config/Private.arm64.exports
index f767c86c..2eef47de 100644
--- a/config/Private.arm64.exports
+++ b/config/Private.arm64.exports
@@ -90,82 +90,17 @@ _mach_bridge_recv_timestamps
 _mach_bridge_init_timestamp
 _mach_bridge_set_params
 _vfs_resolver_result
-_pmap_iommu_init
-_pmap_iommu_iovmalloc
-_pmap_iommu_map
-_pmap_iommu_unmap
-_pmap_iommu_iovmfree
-_pmap_iommu_ioctl
-_pmap_iommu_alloc_contiguous_pages
-_pmap_acquire_iommu_scratch_area
-_pmap_alloc_iommu_page
-_pmap_free_iommu_page
-_pmap_create_aligned_buffer
-_pmap_destroy_aligned_buffer
-_pmap_ref_iommu_table
-_pmap_register_iommu_mapping
-_pmap_release_iommu_scratch_area
-_pmap_unregister_iommu_mapping
 _pmap_cache_attributes
-_nvme_ppl_get_desc
-_sart_get_desc
-_t8020dart_get_desc
-_t8020dart_vo_tte
-_t6000dart_get_desc
-_t6000dart_vo_tte
-_t8110dart_get_desc
-_t8110dart_max_translation_levels
-_t8110dart_vo_tte
-_t8110dart_vo_tt_index
-_sptm_gen3dart_init
-_sptm_gen3dart_map_table
-_sptm_gen3dart_unmap_table
-_sptm_gen3dart_map
-_sptm_gen3dart_unmap
-_sptm_gen3dart_powerdown
-_sptm_gen3dart_powerup
-_sptm_gen3dart_disable_translation
-_sptm_gen3dart_enable_translation
-_sptm_gen3dart_clear_err
-_sptm_gen3dart_clear_perf_interrupts
-_sptm_gen3dart_clear_all_interrupts
-_sptm_gen3dart_disable_all_interrupts
-_sptm_gen3dart_clear_exception
-_sptm_gen3dart_query_tlb
-_sptm_gen3dart_set_smmu_window
-_sptm_gen3dart_read_smmu_stt_index
-_sptm_gen3dart_clamp_tlimits
-_sptm_gen3dart_test_translation
 _ml_amcc_error_inject_register
 _ml_dcs_error_inject_register
 _ml_mem_fault_report_enable_register
 _ml_mcache_flush_callback_register
-_uat_get_desc
 _set_invalidate_hmac_function
 _PE_panic_debugging_enabled
 _register_additional_panic_data_buffer
 _apply_func_phys
 _Switch_context
 _gT1Sz
-__ZN26IOUnifiedAddressTranslator10gMetaClassE
-__ZN26IOUnifiedAddressTranslator10superClassE
-__ZN26IOUnifiedAddressTranslator17getPageTableEntryEy
-__ZN26IOUnifiedAddressTranslator18setClientContextIDEjb
-__ZN26IOUnifiedAddressTranslator21removeClientContextIDEv
-__ZN26IOUnifiedAddressTranslator19isPageFaultExpectedEyj
-__ZN26IOUnifiedAddressTranslator7getModeEv
-__ZN26IOUnifiedAddressTranslator29getFirmwareAddressSpaceHandleEv
-__ZN26IOUnifiedAddressTranslator22registerTaskForServiceEP4taskP9IOService
-__ZN26IOUnifiedAddressTranslator23createMappingInApertureEjP18IOMemoryDescriptorjym
-__ZN26IOUnifiedAddressTranslator23getTotalPageTableMemoryEv
-__ZN26IOUnifiedAddressTranslator3mapEP11IOMemoryMapj
-__ZN26IOUnifiedAddressTranslator5doMapEP18IOMemoryDescriptoryyj
-__ZN26IOUnifiedAddressTranslator5unmapEP11IOMemoryMap
-__ZN26IOUnifiedAddressTranslator7doUnmapEP18IOMemoryDescriptoryy
-__ZN26IOUnifiedAddressTranslator8taskDiedEv
-__ZN26IOUnifiedAddressTranslator12commitUnmapsEv
-__ZN26IOUnifiedAddressTranslator14prepareFWUnmapEyy
-__ZTV26IOUnifiedAddressTranslator
 __ZN24IOBufferMemoryDescriptor18initWithGuardPagesEP4taskjy
 __ZN24IOBufferMemoryDescriptor20inTaskWithGuardPagesEP4taskjy
 __ZN27IOGuardPageMemoryDescriptor12initWithSizeEm
diff --git a/iokit/Kernel/IOKitKernelInternal.h b/iokit/Kernel/IOKitKernelInternal.h
index b33a663d..99c5a808 100644
--- a/iokit/Kernel/IOKitKernelInternal.h
+++ b/iokit/Kernel/IOKitKernelInternal.h
@@ -241,6 +241,10 @@ extern const  OSSymbol * gAKSGetKey;
 void IOScreenLockTimeUpdate(clock_sec_t secs);
 
 void     IOCPUInitialize(void);
+#if defined(__arm64__) && !USE_APPLEARMSMP
+/* iokit/Kernel/arm/OSSARMCPU.cpp - stand-in for the AppleARMCPU kext. */
+void     oss_arm_cpu_initialize(void);
+#endif
 IOReturn IOInstallServicePlatformActions(IOService * service);
 IOReturn IOInstallServiceSleepPlatformActions(IOService * service);
 IOReturn IORemoveServicePlatformActions(IOService * service);
diff --git a/osfmk/arm/commpage/commpage_asm.s b/osfmk/arm/commpage/commpage_asm.s
index 4a74130b..98ae07c7 100644
--- a/osfmk/arm/commpage/commpage_asm.s
+++ b/osfmk/arm/commpage/commpage_asm.s
@@ -275,7 +275,14 @@ _pfz_trylock_and_enqueue:
 	mov		w11, #1			 // locked value = w11 = 1
 
 	// Try to grab the lock
-	casa	w10, w11, [x3]	 // Atomic CAS with acquire barrier
+1:
+	ldaxr	w9, [x3]		 // Load old value (acquire)
+	cmp		w9, w10			 // Compare against expected (0)
+	b.ne	2f
+	stxr	w12, w11, [x3]	 // Conditionally store new value
+	cbnz	w12, 1b
+2:
+	mov		w10, w9			 // Ws := old memory value, matching casa semantics
 	cbz		w10, Ltrylock_enqueue_success
 
 	mov		x0, #-1			// Failed
@@ -325,7 +332,14 @@ _pfz_trylock_and_dequeue:
 	mov		w10, wzr		 // unlock value = w10 = 0
 	mov		w11, #1			 // locked value = w11 = 1
 
-	casa	w10, w11, [x2]	 // Atomic CAS with acquire barrier
+1:
+	ldaxr	w9, [x2]		 // Load old value (acquire)
+	cmp		w9, w10			 // Compare against expected (0)
+	b.ne	2f
+	stxr	w12, w11, [x2]	 // Conditionally store new value
+	cbnz	w12, 1b
+2:
+	mov		w10, w9			 // Ws := old memory value, matching casa semantics
 	cbz		w10, Ltrylock_dequeue_success
 
 	mov		x0, #-1			// Failed
diff --git a/osfmk/arm/pmap/pmap.c b/osfmk/arm/pmap/pmap.c
index 19e7e67a..174572b9 100644
--- a/osfmk/arm/pmap/pmap.c
+++ b/osfmk/arm/pmap/pmap.c
@@ -6180,6 +6180,8 @@ pmap_construct_pte(
 	return pte;
 }
 
+static void pmap_phys_write_disable(vm_address_t va);
+
 MARK_AS_PMAP_TEXT kern_return_t
 pmap_enter_options_internal(
 	pmap_t pmap,
@@ -11055,7 +11057,11 @@ pmap_batch_set_cache_attributes_internal(
 				panic("%s: page is not managed; addr: 0x%016llx", __func__, paddr);
 			}
 
+#if defined(APPLE_ARM64_ARCH_FAMILY) && !APPLEVIRTUALPLATFORM
 			CleanPoC_DcacheRegion_Force_nopreempt_nohid(phystokv(paddr), PAGE_SIZE);
+#else
+			CleanPoC_DcacheRegion_Force_nopreempt(phystokv(paddr), PAGE_SIZE);
+#endif
 
 			page_index++;
 
@@ -12234,6 +12240,23 @@ pmap_unpin_kernel_pages(vm_offset_t kva __unused, size_t nbytes __unused)
 {
 }
 
+/*
+ * Without XNU_MONITOR there is no separate secure trampoline enforcing
+ * code-signing page lockdown, so these are no-ops: page protection is
+ * already whatever the (non-secure) pmap set it to.
+ */
+static void __unused
+pmap_ppl_lockdown_pages(vm_address_t kva __unused, vm_size_t size __unused,
+    uint64_t lockdown_flag __unused, bool ppl_writable __unused)
+{
+}
+
+static void __unused
+pmap_ppl_unlockdown_pages(vm_address_t kva __unused, vm_size_t size __unused,
+    uint64_t lockdown_flag __unused, bool ppl_writable __unused)
+{
+}
+
 #endif /* !XNU_MONITOR */
 
 
diff --git a/osfmk/arm/pmap/pmap.h b/osfmk/arm/pmap/pmap.h
index f9e841f8..2b7552f4 100644
--- a/osfmk/arm/pmap/pmap.h
+++ b/osfmk/arm/pmap/pmap.h
@@ -859,9 +859,9 @@ extern lck_grp_t pmap_lck_grp;
 extern lck_attr_t pmap_lck_rw_attr;
 
 extern void CleanPoC_DcacheRegion_Force_nopreempt_nohid(vm_offset_t va, size_t length);
+extern void CleanPoC_DcacheRegion_Force_nopreempt(vm_offset_t va, size_t length);
 
 #if XNU_MONITOR
-extern void CleanPoC_DcacheRegion_Force_nopreempt(vm_offset_t va, size_t length);
 #define pmap_force_dcache_clean(va, sz) CleanPoC_DcacheRegion_Force_nopreempt(va, sz)
 #define pmap_simple_lock(l)             simple_lock_nopreempt(l, &pmap_lck_grp)
 #define pmap_simple_unlock(l)           simple_unlock_nopreempt(l)
diff --git a/osfmk/arm/pmap/pmap_data.h b/osfmk/arm/pmap/pmap_data.h
index f66dfba3..e4fb72a7 100644
--- a/osfmk/arm/pmap/pmap_data.h
+++ b/osfmk/arm/pmap/pmap_data.h
@@ -478,6 +478,26 @@ pvh_ptep_is_iommu(const pt_entry_t *ptep)
 #endif /* PVH_FLAG_IOMMU */
 }
 
+/**
+ * Given a page table entry pointer retrieved from the pv_head_table that is
+ * known to be an IOMMU mapping (pvh_ptep_is_iommu() returned true), return
+ * an opaque identifier for the owning IOMMU. This is diagnostic-only (used
+ * to format panic() messages) and carries no control-flow significance.
+ *
+ * @param ptep A PTE pointer obtained from the pv_head_table.
+ *
+ * @return An opaque, non-dereferencable identifier suitable for %p.
+ */
+static inline const void *
+ptep_get_iommu(const pt_entry_t *ptep)
+{
+#ifdef PVH_FLAG_IOMMU
+	return (const void *)((vm_offset_t)ptep & ~PVH_FLAG_IOMMU);
+#else /* PVH_FLAG_IOMMU */
+	return ptep;
+#endif /* PVH_FLAG_IOMMU */
+}
+
 /**
  * Sometimes the PTE pointers retrieved from the pv_head_table (from an entry of
  * type PVH_TYPE_PTEP or PVH_TYPE_PVEP) contain flags themselves. This function
diff --git a/osfmk/arm/pmap/pmap_misc.c b/osfmk/arm/pmap/pmap_misc.c
index dacf3b5f..2c48020c 100644
--- a/osfmk/arm/pmap/pmap_misc.c
+++ b/osfmk/arm/pmap/pmap_misc.c
@@ -51,3 +51,11 @@ pmap_abandon_measurement(void)
 	pmap_interrupts_restore(istate);
 #endif /* SCHED_HYGIENE_DEBUG */
 }
+
+#if SCHED_HYGIENE_DEBUG && (DEBUG || DEVELOPMENT)
+bool
+pmap_pending_preemption(void)
+{
+	return _pmap_pending_preemption_real();
+}
+#endif /* SCHED_HYGIENE_DEBUG && (DEBUG || DEVELOPMENT) */
diff --git a/osfmk/arm64/amcc_rorgn_common.c b/osfmk/arm64/amcc_rorgn_common.c
new file mode 100644
index 00000000..d113202e
--- /dev/null
+++ b/osfmk/arm64/amcc_rorgn_common.c
@@ -0,0 +1,14 @@
+/*
+ * amcc_rorgn_common.c is not present in this open-source xnu drop (the
+ * AMCC read-only-region controller is Apple-silicon-specific hardware).
+ * Its public API in amcc_rorgn.h is entirely gated behind
+ * KERNEL_INTEGRITY_KTRR / KERNEL_INTEGRITY_CTRR / KERNEL_INTEGRITY_PV_CTRR,
+ * none of which this board defines (no AMCC controller, no CTRR), so a
+ * faithful implementation for this board is an empty translation unit.
+ */
+
+#include <arm64/amcc_rorgn.h>
+
+#if defined(KERNEL_INTEGRITY_KTRR) || defined(KERNEL_INTEGRITY_CTRR) || defined(KERNEL_INTEGRITY_PV_CTRR)
+#error "amcc_rorgn_common.c stub used on a board that requested AMCC/CTRR kernel integrity enforcement"
+#endif
diff --git a/osfmk/arm64/amcc_rorgn_ppl.c b/osfmk/arm64/amcc_rorgn_ppl.c
new file mode 100644
index 00000000..63866d86
--- /dev/null
+++ b/osfmk/arm64/amcc_rorgn_ppl.c
@@ -0,0 +1,14 @@
+/*
+ * amcc_rorgn_ppl.c is not present in this open-source xnu drop. Its public
+ * API is entirely gated behind KERNEL_INTEGRITY_KTRR / KERNEL_INTEGRITY_CTRR
+ * / KERNEL_INTEGRITY_PV_CTRR, none of which this board defines (no AMCC
+ * controller, no CTRR - config_pmap_ppl is used here only to get the PPL
+ * pmap bootstrap path, not AMCC-backed kernel text protection), so a
+ * faithful implementation for this board is an empty translation unit.
+ */
+
+#include <arm64/amcc_rorgn.h>
+
+#if defined(KERNEL_INTEGRITY_KTRR) || defined(KERNEL_INTEGRITY_CTRR) || defined(KERNEL_INTEGRITY_PV_CTRR)
+#error "amcc_rorgn_ppl.c stub used on a board that requested AMCC/CTRR kernel integrity enforcement"
+#endif
diff --git a/osfmk/arm64/amcc_rorgn_ppl_amcc.c b/osfmk/arm64/amcc_rorgn_ppl_amcc.c
new file mode 100644
index 00000000..9a0f7a37
--- /dev/null
+++ b/osfmk/arm64/amcc_rorgn_ppl_amcc.c
@@ -0,0 +1,13 @@
+/*
+ * amcc_rorgn_ppl_amcc.c is not present in this open-source xnu drop. Its
+ * public API is entirely gated behind KERNEL_INTEGRITY_KTRR /
+ * KERNEL_INTEGRITY_CTRR / KERNEL_INTEGRITY_PV_CTRR, none of which this
+ * board defines (no AMCC controller, no CTRR), so a faithful implementation
+ * for this board is an empty translation unit.
+ */
+
+#include <arm64/amcc_rorgn.h>
+
+#if defined(KERNEL_INTEGRITY_KTRR) || defined(KERNEL_INTEGRITY_CTRR) || defined(KERNEL_INTEGRITY_PV_CTRR)
+#error "amcc_rorgn_ppl_amcc.c stub used on a board that requested AMCC/CTRR kernel integrity enforcement"
+#endif
diff --git a/osfmk/arm64/amcc_rorgn_pv_ctrr.c b/osfmk/arm64/amcc_rorgn_pv_ctrr.c
new file mode 100644
index 00000000..08202f6c
--- /dev/null
+++ b/osfmk/arm64/amcc_rorgn_pv_ctrr.c
@@ -0,0 +1,12 @@
+/*
+ * amcc_rorgn_pv_ctrr.c is not present in this open-source xnu drop. Its
+ * public API is entirely gated behind KERNEL_INTEGRITY_PV_CTRR, which this
+ * board does not define (no paravirtualized CTRR hardware), so a faithful
+ * implementation for this board is an empty translation unit.
+ */
+
+#include <arm64/amcc_rorgn.h>
+
+#if defined(KERNEL_INTEGRITY_PV_CTRR)
+#error "amcc_rorgn_pv_ctrr.c stub used on a board that requested paravirtualized CTRR"
+#endif
diff --git a/osfmk/arm64/arm_vm_init.c b/osfmk/arm64/arm_vm_init.c
index 756af521..2ac01df5 100644
--- a/osfmk/arm64/arm_vm_init.c
+++ b/osfmk/arm64/arm_vm_init.c
@@ -2164,7 +2164,9 @@ arm_vm_init(uint64_t memory_size, boot_args * args)
 	arm_vm_physmap_init(args);
 	set_mmu_ttb_alternate(cpu_ttep & TTBR_BADDR_MASK);
 
+#if !NO_MONITOR
 	ml_enable_monitor();
+#endif /* !NO_MONITOR */
 
 	set_mmu_ttb(invalid_ttep & TTBR_BADDR_MASK);
 
diff --git a/osfmk/arm64/caches_asm.s b/osfmk/arm64/caches_asm.s
index 1588830a..4ae042e0 100644
--- a/osfmk/arm64/caches_asm.s
+++ b/osfmk/arm64/caches_asm.s
@@ -354,6 +354,27 @@ LEXT(CleanPoC_DcacheRegion_Force)
 	b		EXT(CleanPoC_DcacheRegion_internal)
 #endif /* APPLE_ARM64_ARCH_FAMILY */
 
+/*
+ *	void CleanPoC_DcacheRegion_Force_nopreempt(vm_offset_t va, size_t length)
+ *
+ *		Same as CleanPoC_DcacheRegion_Force, but callable when preemption
+ *		is already disabled by the caller. On APPLE_ARM64_ARCH_FAMILY
+ *		this is the (disable/enable-preemption-free) callee of
+ *		CleanPoC_DcacheRegion_Force above; non-Apple-silicon platforms
+ *		have no such split fast path, so this open-source drop never
+ *		defines it for them even though osfmk/arm/pmap/pmap.c calls it
+ *		directly on every architecture. The portable
+ *		CleanPoC_DcacheRegion_internal path doesn't touch preemption
+ *		state either way, so it serves both variants identically.
+ */
+	.text
+	.align 2
+#if !defined(APPLE_ARM64_ARCH_FAMILY)
+	.globl EXT(CleanPoC_DcacheRegion_Force_nopreempt)
+LEXT(CleanPoC_DcacheRegion_Force_nopreempt)
+	ARM64_PROLOG
+	b		EXT(CleanPoC_DcacheRegion_internal)
+#endif /* !APPLE_ARM64_ARCH_FAMILY */
 /*
  *	void FlushPoC_Dcache(void)
  *
diff --git a/osfmk/arm64/corecrypto/arm64_isa_compatibility.h b/osfmk/arm64/corecrypto/arm64_isa_compatibility.h
index 8800687f..8e34ecec 100644
--- a/osfmk/arm64/corecrypto/arm64_isa_compatibility.h
+++ b/osfmk/arm64/corecrypto/arm64_isa_compatibility.h
@@ -37,7 +37,7 @@
 // #include <Availability.h>
 #include <sys/cdefs.h>
 
-#if defined(__clang__) && ((defined(__apple_build_version__) && __apple_build_version__ > 5010000))
+#if defined(__clang__) && ((defined(__apple_build_version__) && __apple_build_version__ > 5010000) || !defined(__apple_build_version__))
 #define __USES_V_CRYPTO_INTRINSICS 1
 #else
 #define __USES_V_CRYPTO_INTRINSICS 0
diff --git a/osfmk/arm64/machine_routines.c b/osfmk/arm64/machine_routines.c
index 07143d53..ed6e4d52 100644
--- a/osfmk/arm64/machine_routines.c
+++ b/osfmk/arm64/machine_routines.c
@@ -3384,7 +3384,14 @@ __startup_func
 static void
 ml_unsafe_kernel_text_init(void)
 {
-	/* Grab the values written by iBoot. */
+	/*
+	 * iBoot writes /chosen/kernel-ctrr-to-be-enabled. QEMU/OSS firmware
+	 * does not, and DEVELOPMENT asserts that this flag was initialized
+	 * before the first ml_unsafe_kernel_text() call (sdt_init).
+	 * Missing property means CTRR will not lock text.
+	 */
+	_unsafe_kernel_text_initialized = true;
+	_unsafe_kernel_text = true;
 
 	DTEntry         entry;
 	const void      *value;
@@ -3392,7 +3399,6 @@ ml_unsafe_kernel_text_init(void)
 	if (SecureDTLookupEntry(0, "/chosen", &entry) == kSuccess &&
 	    SecureDTGetProperty(entry, "kernel-ctrr-to-be-enabled", &value, &size) == kSuccess &&
 	    size == sizeof(int)) {
-		_unsafe_kernel_text_initialized = true;
 		_unsafe_kernel_text = (0 == *(const int *)value);
 	}
 }
@@ -3406,3 +3412,15 @@ ml_unsafe_kernel_text(void)
 	return false;
 }
 #endif /* DEVELOPMENT || DEBUG || CONFIG_DTRACE || CONFIG_CSR_FROM_DT */
+
+/*
+ * This SoC (Amlogic S905D2 / G12A) has no Apple-silicon eFuse array to
+ * query. Report "production fused" unconditionally, which disables the
+ * engineering/dev-fused-only diagnostic paths that key off this - the
+ * conservative choice given there is no real fuse state to read.
+ */
+boolean_t
+ml_device_is_prod_fused(void)
+{
+	return TRUE;
+}
diff --git a/osfmk/arm64/ppl/sart.c b/osfmk/arm64/ppl/sart.c
new file mode 100644
index 00000000..3c0be65a
--- /dev/null
+++ b/osfmk/arm64/ppl/sart.c
@@ -0,0 +1,21 @@
+/*
+ * sart.c is not present in this open-source xnu drop. SART (Secure Address
+ * Range Table) is Apple-silicon-specific DMA range-protection hardware.
+ * This board (Amlogic G12A) has none: sart_bootstrap() looks for an
+ * "sart" node in the boot device tree (exactly like a real implementation
+ * would), finds none since no such node is ever populated for this board,
+ * and returns.
+ */
+
+#include <arm64/ppl/sart.h>
+#include <pexpert/device_tree.h>
+
+void
+sart_bootstrap(void)
+{
+	DTEntry sart_node;
+
+	if (SecureDTLookupEntry(NULL, "/arm-io/sart", &sart_node) != kSuccess) {
+		return;
+	}
+}
diff --git a/osfmk/arm64/ppl/sart.h b/osfmk/arm64/ppl/sart.h
new file mode 100644
index 00000000..2ef1795f
--- /dev/null
+++ b/osfmk/arm64/ppl/sart.h
@@ -0,0 +1,19 @@
+/*
+ * sart.h is not present in this open-source xnu drop. SART (Secure Address
+ * Range Table) is Apple-silicon-specific DMA range-protection hardware;
+ * this board (Amlogic G12A) has none, so sart_bootstrap() is implemented
+ * as a device-tree lookup that finds no "sart" node and returns.
+ */
+
+#ifndef _ARM64_PPL_SART_H_
+#define _ARM64_PPL_SART_H_
+
+#include <sys/cdefs.h>
+
+__BEGIN_DECLS
+
+void sart_bootstrap(void);
+
+__END_DECLS
+
+#endif /* _ARM64_PPL_SART_H_ */
diff --git a/osfmk/arm64/ppl/uat.h b/osfmk/arm64/ppl/uat.h
new file mode 100644
index 00000000..91d03fc1
--- /dev/null
+++ b/osfmk/arm64/ppl/uat.h
@@ -0,0 +1,11 @@
+/*
+ * uat.h is not present in this open-source xnu drop. UAT is an
+ * Apple-silicon-specific IOMMU/DART mechanism; this board (Amlogic G12A)
+ * has none, and nothing in the sources built for this board calls into
+ * it, so this header is intentionally empty.
+ */
+
+#ifndef _ARM64_PPL_UAT_H_
+#define _ARM64_PPL_UAT_H_
+
+#endif /* _ARM64_PPL_UAT_H_ */
diff --git a/osfmk/arm64/proc_reg.h b/osfmk/arm64/proc_reg.h
index f4d148bd..b2349594 100644
--- a/osfmk/arm64/proc_reg.h
+++ b/osfmk/arm64/proc_reg.h
@@ -251,6 +251,27 @@
 /* D-Cache. */
 #define MMU_CLINE   6
 
+#elif defined (SUPERBIRD)
+
+/* Cortex-A53: 32KB 2-way I-cache, 32KB 4-way D-cache, 64-byte lines. */
+#define MMU_I_CLINE 6
+
+#define MMU_CLINE   6
+
+#elif defined (QEMU)
+
+/* Cortex-A53 (QEMU virt -cpu cortex-a53): 32KB 2-way I-cache, 32KB 4-way
+ * D-cache, 64-byte lines - same geometry as SUPERBIRD's real A53. */
+#define MMU_I_CLINE 6
+
+#define MMU_CLINE   6
+#elif defined (APPLECYCLONE)
+
+/* Cyclone (A7): 64KB 4-way I-cache, 64KB 4-way D-cache, 64-byte lines. */
+#define MMU_I_CLINE 6
+
+#define MMU_CLINE   6
+
 #else
 #error processor not supported
 #endif
@@ -1973,7 +1994,6 @@
 #define ARM_4K_TT_L3_OFFMASK    0x0000000000000fffULL /* offset within L3 PTE */
 #define ARM_4K_TT_L3_SHIFT      12                    /* page descriptor shift */
 #define ARM_4K_TT_L3_INDEX_MASK 0x00000000001ff000ULL /* mask for page descriptor index */
-
 #ifdef __ARM_16K_PG__
 
 /* Native L0 defines */
diff --git a/osfmk/arm64/sleh.c b/osfmk/arm64/sleh.c
index 0554aa80..ab7f4ed9 100644
--- a/osfmk/arm64/sleh.c
+++ b/osfmk/arm64/sleh.c
@@ -674,15 +674,11 @@ sleh_synchronous_sp1(arm_context_t *context, uint64_t esr, vm_offset_t far __unu
 	switch (class) {
 	case ESR_EC_UNCATEGORIZED:
 	{
-#if (DEVELOPMENT || DEBUG)
 		uint32_t instr = *((uint32_t*)get_saved_state_pc(state));
 		if (IS_ARM_GDB_TRAP(instr)) {
 			DebuggerCall(EXC_BREAKPOINT, state);
 		}
 		OS_FALLTHROUGH; // panic if we return from the debugger
-#else
-		panic_with_thread_kernel_state("Unexpected debugger trap while SP1 selected", state);
-#endif /* (DEVELOPMENT || DEBUG) */
 	}
 	default:
 		panic_with_thread_kernel_state("Synchronous exception taken while SP1 selected", state);
diff --git a/osfmk/arm64/tunables/tunables.s b/osfmk/arm64/tunables/tunables.s
new file mode 100644
index 00000000..cea997aa
--- /dev/null
+++ b/osfmk/arm64/tunables/tunables.s
@@ -0,0 +1,13 @@
+/*
+ * arm64/tunables/tunables.s is not present in this open-source xnu drop:
+ * it dispatches on MIDR_EL1 to apply per-Apple-SoC IMPLEMENTATION DEFINED
+ * errata/tunable register writes. This board (Cortex-A53, Amlogic G12A)
+ * has no Apple silicon and none of those tunables apply, so APPLY_TUNABLES
+ * is a no-op here.
+ *
+ * @param $0 MIDR_EL1 value (unused)
+ * @param $1 scratch register (unused)
+ * @param $2 scratch register (unused)
+ */
+.macro APPLY_TUNABLES
+.endmacro
diff --git a/osfmk/conf/files.arm64 b/osfmk/conf/files.arm64
index 878b5154..ba3fcd6e 100644
--- a/osfmk/conf/files.arm64
+++ b/osfmk/conf/files.arm64
@@ -13,6 +13,7 @@ osfmk/arm/pmap/pmap_data.c		optional nos_arm_pmap
 osfmk/arm/pmap/pmap_iommu.c		optional nos_arm_pmap
 osfmk/arm/pmap/pmap_misc.c		optional nos_arm_pmap
 osfmk/arm/pmap/pmap_ppl_interface.c	optional nos_arm_pmap
+osfmk/arm64/ppl/sart.c			optional nos_arm_pmap
 
 osfmk/arm64/sptm/pmap/pmap.c		optional nos_arm_asm config_sptm
 osfmk/arm64/sptm/pmap/pmap_cs.c		optional nos_arm_asm config_sptm
diff --git a/osfmk/kern/zalloc.c b/osfmk/kern/zalloc.c
index 9edf8c5c..c82b23ad 100644
--- a/osfmk/kern/zalloc.c
+++ b/osfmk/kern/zalloc.c
@@ -167,7 +167,18 @@ ZONE_DEFINE_TYPE(percpu_u64_zone, "percpu.64", uint64_t,
 #define ZONE_MAX_CHUNK_ALLOC_NUM        (10)
 #endif /* ZSECURITY_CONFIG(SAD_FENG_SHUI) */
 
-#if   XNU_PLATFORM_MacOSX
+#if defined(ARM64_BOARD_CONFIG_QEMU)
+/*
+ * QEMU virt is compiled with XNU_PLATFORM_MacOSX, whose default zone
+ * map is 128GB of VA. The 4K-page ARM64 kernel window is only 112GB
+ * (VM_MIN 0xffffffe000000000 .. VM_MAX 0xfffffffbffffffff) and after
+ * pmap_virtual_region reservations the remaining hole is ~28GB.
+ * kmem_add_extra_claims() then underflows free_size and panics in
+ * kmem_range_init with a nonsense kmem_ptr_range_0 size.
+ */
+#define ZONE_MAP_MAX            (2ULL << 30)
+#define ZONE_MAP_VA_SIZE        (8ULL << 30)
+#elif   XNU_PLATFORM_MacOSX
 #define ZONE_MAP_MAX            (32ULL << 30)
 #define ZONE_MAP_VA_SIZE        (128ULL << 30)
 #else
diff --git a/osfmk/mach/arm/vm_param.h b/osfmk/mach/arm/vm_param.h
index 8d11b54e..5fb3c3ee 100644
--- a/osfmk/mach/arm/vm_param.h
+++ b/osfmk/mach/arm/vm_param.h
@@ -134,8 +134,22 @@ extern int PAGE_SHIFT_CONST;
 /* system-wide values */
 #define MACH_VM_MIN_ADDRESS_RAW 0x0ULL
 #if defined(XNU_PLATFORM_MacOSX) || defined(XNU_PLATFORM_DriverKit)
+#if __ARM_16K_PG__
 #define MACH_VM_MAX_ADDRESS_RAW 0x00007FFFFE000000ULL
 #else
+/*
+ * 4K-page macOS-flavored boards (SUPERBIRD/IPAD41/QEMU - no real Apple
+ * silicon macOS target is ever 4K) can't use the full 47-bit macOS
+ * ceiling here: osfmk/arm/pmap/pmap.c's commpage-nesting static_assert
+ * for ARM_PGSHIFT==12 requires _COMM_PAGE64_BASE_ADDRESS to sit above
+ * MACH_VM_MAX_ADDRESS, which only holds if this stays the smaller
+ * iOS-style ceiling - the 16K branch's macOS-sized ceiling is exempt
+ * from that assert only because 16K macOS nests the commpage inside
+ * userspace as a VM-reserved region instead (see that assert's comment).
+ */
+#define MACH_VM_MAX_ADDRESS_RAW 0x0000000FC0000000ULL
+#endif
+#else
 #define MACH_VM_MAX_ADDRESS_RAW 0x0000000FC0000000ULL
 #endif
 
diff --git a/pexpert/aic/aic.c b/pexpert/aic/aic.c
new file mode 100644
index 00000000..bddc284b
--- /dev/null
+++ b/pexpert/aic/aic.c
@@ -0,0 +1,96 @@
+/*
+ * AICv1 (Apple Interrupt Controller, pre-AIC2) driver skeleton for iPad4,1 / A7 (s5l8960x).
+ */
+#include <stdint.h>
+#include <stddef.h>
+#include <pexpert/pexpert.h>
+#include <pexpert/arm64/gic.h>
+
+#ifndef AIC_PHYS_BASE
+#define AIC_PHYS_BASE 0x20e100000ULL
+#endif
+
+#ifndef AIC_SIZE
+#define AIC_SIZE 0x8000
+#endif
+
+#ifndef AIC_CONFIG
+#define AIC_CONFIG 0x2000
+#endif
+
+#ifndef AIC_MASK_SET
+#define AIC_MASK_SET 0x3c00
+#endif
+
+#ifndef AIC_MASK_CLR
+#define AIC_MASK_CLR 0x3c04
+#endif
+
+#ifndef AIC_EVENT
+#define AIC_EVENT 0x3804
+#endif
+
+#ifndef AIC_SPURIOUS_IRQ
+#define AIC_SPURIOUS_IRQ 192
+#endif
+
+#ifndef MAX_IRQS
+#define MAX_IRQS AIC_SPURIOUS_IRQ
+#endif
+
+extern vm_offset_t ml_io_map(vm_offset_t phys_addr, vm_size_t size);
+extern void ml_install_interrupt_handler(
+    void *nub,
+    int source,
+    void *target,
+    void (*handler)(void *, void *, void *, int),
+    void *ref_con
+);
+
+static vm_offset_t aic_base = 0;
+
+static inline uint32_t mmio_read32(vm_offset_t addr) {
+    return *(volatile uint32_t *)addr;
+}
+
+static inline void mmio_write32(vm_offset_t addr, uint32_t val) {
+    *(volatile uint32_t *)addr = val;
+}
+
+static inline uint32_t aic_read(uint32_t offset) {
+    return mmio_read32(aic_base + offset);
+}
+
+static inline void aic_write(uint32_t offset, uint32_t val) {
+    mmio_write32(aic_base + offset, val);
+}
+
+static void aic_dispatch(void *target, void *refCon, void *nub, int source) {
+    (void)target;
+    (void)refCon;
+    (void)nub;
+    (void)source;
+    while (1) {
+        uint32_t irq = aic_read(AIC_EVENT);
+        if (irq >= AIC_SPURIOUS_IRQ) {
+            break;
+        }
+        aic_write(AIC_EVENT, irq);
+    }
+}
+
+void gic_init(void) {
+    if (aic_base != 0) {
+        return;
+    }
+    aic_base = ml_io_map(AIC_PHYS_BASE, AIC_SIZE);
+    aic_write(AIC_MASK_SET, 0xffffffff);
+    ml_install_interrupt_handler(NULL, 0, NULL, aic_dispatch, NULL);
+}
+
+void gic_enable_interrupt(uint32_t irq) {
+    if (aic_base == 0 || irq >= MAX_IRQS) {
+        return;
+    }
+    aic_write(AIC_MASK_CLR, (1U << (irq % 32)));
+}
diff --git a/pexpert/arm/pe_init.c b/pexpert/arm/pe_init.c
index db1c4a02..ac10bcae 100644
--- a/pexpert/arm/pe_init.c
+++ b/pexpert/arm/pe_init.c
@@ -21,6 +21,7 @@
 #include <os/overflow.h>
 
 #include <pexpert/arm64/board_config.h>
+#include <pexpert/arm64/gic.h>
 
 #if CONFIG_SPTM
 #include <arm64/sptm/sptm.h>
@@ -327,6 +328,11 @@ PE_init_iokit(void)
 	show_progress = FALSE;
 	PE_parse_boot_argn("-progress", &show_progress, sizeof(show_progress));
 #endif /* XNU_TARGET_OS_OSX */
+	/* Headless: the delta loop below is `while (image >= display/2)`
+	 * with display_size 0, which never terminates in RELEASE. */
+	if (PE_state.video.v_width == 0 || PE_state.video.v_height == 0) {
+		show_progress = FALSE;
+	}
 	if (show_progress) {
 		/* Rotation: 0:normal, 1:right 90, 2:left 180, 3:left 90 */
 		switch (PE_state.video.v_rotate) {
@@ -408,18 +414,6 @@ PE_init_iokit(void)
 void
 PE_lockdown_iokit(void)
 {
-	/*
-	 * On arm/arm64 platforms, and especially those that employ KTRR/CTRR,
-	 * machine_lockdown() is treated as a hard security checkpoint, such that
-	 * code which executes prior to lockdown must be minimized and limited only to
-	 * trusted parts of the kernel and specially-entitled kexts.  We therefore
-	 * cannot start the general-purpose IOKit matching process until after lockdown,
-	 * as it may involve execution of untrusted/non-entitled kext code.
-	 * Furthermore, such kext code may process attacker controlled data (e.g.
-	 * network packets), which dramatically increases the potential attack surface
-	 * against a kernel which has not yet enabled the full set of available
-	 * hardware protections.
-	 */
 	zalloc_iokit_lockdown();
 	StartIOKitMatching();
 }
@@ -469,6 +463,7 @@ PE_init_platform(boolean_t vm_initialized, void *args)
 		pe_identify_machine(boot_args_ptr);
 	} else {
 		pe_arm_init_interrupts(args);
+		gic_init();
 	}
 
 	if (!vm_initialized) {
diff --git a/pexpert/arm/pe_kprintf.c b/pexpert/arm/pe_kprintf.c
index 7a4bbd39..89b189d9 100644
--- a/pexpert/arm/pe_kprintf.c
+++ b/pexpert/arm/pe_kprintf.c
@@ -12,6 +12,9 @@
 #include <kern/simple_lock.h>
 #include <os/log_private.h>
 #include <libkern/section_keywords.h>
+#ifdef __arm64__
+#include <pexpert/arm64/board_config.h>
+#endif
 
 /* Globals */
 typedef void (*PE_kputc_t)(char);
@@ -21,9 +24,14 @@ SECURITY_READ_ONLY_LATE(PE_kputc_t) PE_kputc;
  * disable_serial_output disables kprintf() *and* unbuffered panic output
  * except for SPTM based early panic serial output via `sptm_serial_putc()`.
  */
-SECURITY_READ_ONLY_LATE(bool) disable_serial_output = true;
-// disable_kprintf_output only disables kprintf().
-SECURITY_READ_ONLY_LATE(bool) disable_kprintf_output = true;
+#if defined(QEMU)
+	/* QEMU virt has no framebuffer; serial is the only console. */
+	SECURITY_READ_ONLY_LATE(bool) disable_serial_output = false;
+	SECURITY_READ_ONLY_LATE(bool) disable_kprintf_output = false;
+#else
+	SECURITY_READ_ONLY_LATE(bool) disable_serial_output = true;
+	SECURITY_READ_ONLY_LATE(bool) disable_kprintf_output = true;
+#endif
 // disable_iolog_serial_output only disables IOLog, controlled by
 // SERIALMODE_NO_IOLOG.
 SECURITY_READ_ONLY_LATE(bool) disable_iolog_serial_output = false;
@@ -44,6 +52,9 @@ PE_init_kprintf_config(void)
 	if (debug_boot_arg & (DB_KPRT | DB_PRT)) {
 		disable_serial_output = false;
 	}
+	if (debug_boot_arg & DB_KPRT) {
+		disable_kprintf_output = false;
+	}
 
 #if DEBUG
 	disable_kprintf_output = false;
diff --git a/pexpert/conf/files.arm64 b/pexpert/conf/files.arm64
index 9191100c..fe270c5b 100644
--- a/pexpert/conf/files.arm64
+++ b/pexpert/conf/files.arm64
@@ -6,6 +6,9 @@ pexpert/arm/pe_init.c                  standard
 pexpert/arm/pe_kprintf.c               standard
 pexpert/arm/pe_serial.c                standard
 
+pexpert/gic/gic.c                      optional not applecyclone
+pexpert/aic/aic.c                      optional applecyclone
+
 pexpert/arm/hwtrace/hwtrace.c         standard
 pexpert/arm/hwtrace/apt_msg.c         standard
 
diff --git a/pexpert/gic/gic.c b/pexpert/gic/gic.c
new file mode 100644
index 00000000..8c8360ae
--- /dev/null
+++ b/pexpert/gic/gic.c
@@ -0,0 +1,156 @@
+/*
+ * Minimal GIC-400 (GICv2) MMIO driver, shared by every board that uses a
+ * plain GICv2 distributor + banked CPU interface (no ICC_* system registers).
+ */
+#include <stdint.h>
+#include <stddef.h>
+#include <pexpert/pexpert.h>
+#include <pexpert/arm64/gic.h>
+
+#ifndef GIC_SPURIOUS_IRQ
+#define GIC_SPURIOUS_IRQ 1023
+#endif
+
+#ifndef GIC_IRQ_MASK
+#define GIC_IRQ_MASK 0x3ff
+#endif
+
+#ifndef MAX_IRQS
+#define MAX_IRQS 1020
+#endif
+
+#ifndef GICD_CTLR
+#define GICD_CTLR 0x000
+#endif
+
+#ifndef GICD_ISENABLER
+#define GICD_ISENABLER 0x100
+#endif
+
+#ifndef GICC_CTLR
+#define GICC_CTLR 0x000
+#endif
+
+#ifndef GICC_PMR
+#define GICC_PMR 0x004
+#endif
+
+#ifndef GICC_BPR
+#define GICC_BPR 0x008
+#endif
+
+#ifndef GICC_IAR
+#define GICC_IAR 0x00c
+#endif
+
+#ifndef GICC_EOIR
+#define GICC_EOIR 0x010
+#endif
+
+#ifndef GICD_CTLR_ENABLEGRP0
+#define GICD_CTLR_ENABLEGRP0 (1 << 0)
+#endif
+
+#ifndef GICD_CTLR_ENABLEGRP1
+#define GICD_CTLR_ENABLEGRP1 (1 << 1)
+#endif
+
+#ifndef GICC_CTLR_ENABLEGRP0
+#define GICC_CTLR_ENABLEGRP0 (1 << 0)
+#endif
+
+#ifndef GICC_CTLR_ENABLEGRP1
+#define GICC_CTLR_ENABLEGRP1 (1 << 1)
+#endif
+
+extern vm_offset_t ml_io_map(vm_offset_t phys_addr, vm_size_t size);
+extern void ml_install_interrupt_handler(
+    void *nub,
+    int source,
+    void *target,
+    void (*handler)(void *, void *, void *, int),
+    void *ref_con
+);
+extern boolean_t ml_get_timer_pending(void);
+extern void rtclock_intr(unsigned int is_user_context);
+
+/* GICv2 INTID for the ARM generic virtual EL1 timer PPI, as wired up by
+ * QEMU's virt machine device tree (PPI #11 -> GIC INTID 27). */
+#ifndef GIC_VIRT_TIMER_IRQ
+#define GIC_VIRT_TIMER_IRQ 27
+#endif
+
+static vm_offset_t gicd_base = 0;
+static vm_offset_t gicc_base = 0;
+
+static inline uint32_t mmio_read32(vm_offset_t addr) {
+    return *(volatile uint32_t *)addr;
+}
+
+static inline void mmio_write32(vm_offset_t addr, uint32_t val) {
+    *(volatile uint32_t *)addr = val;
+}
+
+static inline uint32_t gicd_read(uint32_t offset) {
+    return mmio_read32(gicd_base + offset);
+}
+
+static inline void gicd_write(uint32_t offset, uint32_t val) {
+    mmio_write32(gicd_base + offset, val);
+}
+
+static inline uint32_t gicc_read(uint32_t offset) {
+    return mmio_read32(gicc_base + offset);
+}
+
+static inline void gicc_write(uint32_t offset, uint32_t val) {
+    mmio_write32(gicc_base + offset, val);
+}
+
+static void gic_dispatch(void *target, void *refCon, void *nub, int source) {
+    (void)target;
+    (void)refCon;
+    (void)nub;
+    (void)source;
+    while (1) {
+        uint32_t iar = gicc_read(GICC_IAR);
+        uint32_t irq = iar & GIC_IRQ_MASK;
+        if (irq >= GIC_SPURIOUS_IRQ) {
+            break;
+        }
+        if (irq == GIC_VIRT_TIMER_IRQ || ml_get_timer_pending()) {
+            rtclock_intr(TRUE);
+        }
+        gicc_write(GICC_EOIR, iar);
+    }
+}
+
+void gic_init(void) {
+    if (gicd_base != 0) {
+        return;
+    }
+#if defined(GICD_PHYS_BASE) && defined(GICC_PHYS_BASE)
+    gicd_base = ml_io_map(GICD_PHYS_BASE, GICD_SIZE);
+    gicc_base = ml_io_map(GICC_PHYS_BASE, GICC_SIZE);
+
+    gicd_write(GICD_CTLR, 0);
+    gicd_write(GICD_CTLR, GICD_CTLR_ENABLEGRP0 | GICD_CTLR_ENABLEGRP1);
+
+    gicc_write(GICC_PMR, 0xff);
+    gicc_write(GICC_BPR, 0);
+    gicc_write(GICC_CTLR, GICC_CTLR_ENABLEGRP0 | GICC_CTLR_ENABLEGRP1);
+
+    gicd_write(GICD_ISENABLER + (GIC_VIRT_TIMER_IRQ / 32) * 4, 1U << (GIC_VIRT_TIMER_IRQ % 32));
+
+    ml_install_interrupt_handler(NULL, 0, NULL, gic_dispatch, NULL);
+#endif
+}
+
+void gic_enable_interrupt(uint32_t irq) {
+    if (gicd_base == 0 || irq >= MAX_IRQS) {
+        return;
+    }
+    uint32_t reg = GICD_ISENABLER + (irq / 32) * 4;
+    uint32_t shift = irq % 32;
+    gicd_write(reg, (1U << shift));
+}
diff --git a/pexpert/pexpert/arm64/IPAD41.h b/pexpert/pexpert/arm64/IPAD41.h
new file mode 100644
index 00000000..ee22d28b
--- /dev/null
+++ b/pexpert/pexpert/arm64/IPAD41.h
@@ -0,0 +1,80 @@
+/*
+ * Board support header for the iPad Air (iPad4,1), Apple A7 / s5l8960x,
+ * dual Cyclone (ARMv8.0-A). Targeted for boot via pongoOS (checkm8/KPF),
+ * which loads the raw Mach-O this build produces.
+ *
+ * A7 predates nearly every Apple-silicon feature this tree guards on:
+ * no PPL/SPTM, no KTRR/AMCC (those arrive with A10), no PAC, no PAN,
+ * no GIC — interrupts are the Samsung-derived AICv1. Like SUPERBIRD.h,
+ * everything gated behind those features must stay off.
+ */
+
+#ifndef _PEXPERT_ARM64_IPAD41_H
+#define _PEXPERT_ARM64_IPAD41_H
+
+#define NO_MONITOR                1
+#define IPAD41                    1
+/* proc_reg.h cache-geometry arm key (Cyclone predates APPLETYPHOON). */
+#define APPLECYCLONE              1
+
+/*
+ * No SEP/Tightbeam/exclaves path bring-up for this port (A7's SEP exists
+ * but is unused here); reuse the non-Apple-silicon escape hatch so the
+ * same source guards apply.
+ */
+#define OSS_HARDWARE               1
+#define NO_ECORE                  1
+
+/*
+ * No PAC/PAN/SME/PAN3 - Cyclone is plain ARMv8.0-A. Deliberately do NOT
+ * define CPU_HAS_APPLE_PAC, HAS_PARAVIRTUALIZED_PAC,
+ * HAS_PARAVIRTUALIZED_CTRR, HAS_ARM_FEAT_SSBS2/SME/SME2/PAN3: none exist
+ * on this CPU and defining them would emit instructions that #UD here.
+ * A7 also predates PPL entirely; the SPTM pmap variant
+ * (osfmk/arm64/sptm/pmap.c) must not be selected for this board — the
+ * kernel config side of that choice is still TODO.
+ */
+
+/*
+ * Deliberately left UNDEFINED, not `#define __ARM_16K_PG__ 0` (A7 is a
+ * 4K-page device) - see pexpert/pexpert/arm64/QEMU.h's comment:
+ * osfmk/arm64/proc_reg.h mixes `#ifdef __ARM_16K_PG__` and
+ * `#if __ARM_16K_PG__` guards, which disagree if this is
+ * defined-but-zero, producing a self-inconsistent TCR_EL1 (16K-style
+ * T0SZ with a 4K TG0) that instruction-aborts at SCTLR_EL1.M-enable
+ * time.
+ */
+#define __ARM_RANGE_TLBI__        0
+
+#define ARM_PARAMETERIZED_PMAP    1
+
+#include <pexpert/arm64/apple_arm64_common.h>
+#undef  BTI_ENFORCED
+#define BTI_ENFORCED 0
+#undef  __ARM64_PMAP_SUBPAGE_L1__
+#undef  __ARM64_PMAP_KERN_SUBPAGE_L1__
+/* Not treated as an Apple-arch-family target here: skips Apple-specific
+ * cache-maintenance helpers this fork does not bring up. */
+#undef  APPLE_ARM64_ARCH_FAMILY
+
+/*
+ * Console: A7 exposes a Samsung-style UART (uart0 @ 0x20a0c0000 per the
+ * iPad4,1 device tree) usable with `serial=3` under pongoOS; no driver
+ * is wired up in this tree yet. pongoOS also hands a framebuffer via
+ * xnu_boot_arguments.video_information for early panic display.
+ */
+#ifndef ASSEMBLER
+#define PLATFORM_PANIC_LOG_DISABLED
+#endif /* ! ASSEMBLER */
+
+/*
+ * Interrupt controller: AICv1 (aic@20e100000 in the iPad4,1 device tree).
+ * There is no GIC at all; do NOT define HAS_GIC_V3 or GICD/GICC_* here.
+ * The driver is pexpert/aic/aic.zig, built by the zig build for this
+ * board and exporting the gic_init/gic_enable_interrupt ABI.
+ */
+#define AIC_PHYS_BASE             0x20e100000ULL
+#define AIC_SIZE                  0x8000
+#define AIC_SPURIOUS_IRQ          192 /* A7 AIC advertises 192 lines */
+
+#endif /* ! _PEXPERT_ARM64_IPAD41_H */
diff --git a/pexpert/pexpert/arm64/Makefile b/pexpert/pexpert/arm64/Makefile
index 75938b56..7ef16aaa 100644
--- a/pexpert/pexpert/arm64/Makefile
+++ b/pexpert/pexpert/arm64/Makefile
@@ -33,7 +33,11 @@ PRIVATE_KERNELFILES = \
 	H15.h \
 	H16.h \
 	H17.h \
-	VMAPPLE.h
+	VMAPPLE.h \
+	QEMU.h \
+	SUPERBIRD.h \
+	IPAD41.h \
+	gic.h
 
 
 # Headers installed into System.framework/PrivateHeaders (userspace internal SDK only).
diff --git a/pexpert/pexpert/arm64/QEMU.h b/pexpert/pexpert/arm64/QEMU.h
new file mode 100644
index 00000000..879568eb
--- /dev/null
+++ b/pexpert/pexpert/arm64/QEMU.h
@@ -0,0 +1,110 @@
+/*
+ * Board support header for QEMU's aarch64 "virt" machine
+ * (`qemu-system-aarch64 -M virt -cpu cortex-a53`), booted the same way as
+ * SUPERBIRD: u-boot's generic `bootxnu` command (../u-boot/cmd/bootxnu.c -
+ * board-agnostic, just parses the Mach-O this build produces and jumps
+ * into it with a boot_args struct) run from u-boot built with
+ * qemu_arm64_defconfig.
+ *
+ * `-cpu cortex-a53` is plain ARMv8.0-A, same silicon class as SUPERBIRD:
+ * no PAC, no PPL/SPTM, no KTRR/AMCC, no Apple Interrupt Controller.
+ * Everything gated behind those features must stay off, same as
+ * SUPERBIRD.h.
+ */
+
+#ifndef _PEXPERT_ARM64_QEMU_H
+#define _PEXPERT_ARM64_QEMU_H
+
+#define NO_MONITOR                1
+#define QEMU                      1
+/*
+ * Generic flag for non-Apple-silicon boards: no Tightbeam/exclaves (no
+ * SEP/secure-enclave coprocessor to talk to), no per-SoC blob.
+ */
+#define OSS_HARDWARE               1
+#define NO_ECORE                  1
+
+/*
+ * No Apple Silicon PAC/CTRR/SME/PAN3 - `-cpu cortex-a53` is plain
+ * ARMv8.0-A. Deliberately do NOT define CPU_HAS_APPLE_PAC,
+ * HAS_PARAVIRTUALIZED_PAC, HAS_PARAVIRTUALIZED_CTRR,
+ * HAS_ARM_FEAT_SSBS2/SME/SME2/PAN3: none exist on this CPU model and
+ * defining them would emit instructions that #UD here.
+ */
+
+/*
+ * Deliberately left UNDEFINED, not `#define __ARM_16K_PG__ 0`: this is a
+ * 4K-page device, but osfmk/arm64/proc_reg.h guards T0SZ_BOOT with
+ * `#ifdef __ARM_16K_PG__` (true if *defined*, even to 0) while it guards
+ * TCR_TG0_GRANULE_SIZE/ARM_TT_L2_SIZE/ARM_PGSHIFT with
+ * `#if __ARM_16K_PG__` (true only if *nonzero*). Defining this to 0 trips
+ * the former into the 16K-page T0SZ (17, expecting a 4-level walk) while
+ * everything else takes the 4K-page branch (T0SZ 25, 3-level
+ * start.s bootstrap tables + TG0=4K) - a self-inconsistent TCR_EL1 that
+ * instruction-aborts the moment SCTLR_EL1.M is set. Leaving it undefined
+ * makes both guards agree on the 4K-page branch, matching real Apple
+ * boards, which never define this macro for 4K-page devices.
+ */
+#define __ARM_RANGE_TLBI__        0
+
+#define ARM_PARAMETERIZED_PMAP    1
+
+#include <pexpert/arm64/apple_arm64_common.h>
+#undef  BTI_ENFORCED
+#define BTI_ENFORCED 0
+#undef  __ARM64_PMAP_SUBPAGE_L1__
+#undef  __ARM64_PMAP_KERN_SUBPAGE_L1__
+/* Crypto extensions unconfirmed for QEMU's cortex-a53 model - use generic C. */
+#undef  __ARM_V8_CRYPTO_EXTENSIONS__
+/* This is not Apple silicon: no coherent-fabric assumptions, no Apple
+ * cache-maintenance helpers (nopreempt/mva_ops), no eng-fused sysctl. */
+#undef  APPLE_ARM64_ARCH_FAMILY
+
+/*
+ * Console: PL011 UART at 0x09000000 (`-M virt`'s pl011@9000000). The
+ * driver already exists in pexpert/arm/pe_serial.c, gated on
+ * PL011_UART; it just needed a board that enables it and an AFDT
+ * describing the device (arm-io/defaults/pl011 nodes - see
+ * tools/qemu-boot/boot.zig's AFDT builder, the only source of the
+ * device tree this board boots with, and boot_args.command_line's
+ * `serial-device-name=uart0` selecting it via pe_serial.c's
+ * get_serial_device_phandle()).
+ */
+#define PL011_UART 1
+
+#ifndef ASSEMBLER
+#define PLATFORM_PANIC_LOG_DISABLED
+#endif /* ! ASSEMBLER */
+
+/*
+ * GICv2, `-M virt`'s default (`gic-version=2`; `gic-version=3` would need
+ * the ICC_*-system-register driver in pexpert/arm/pe_fiq.c instead, not
+ * this MMIO one):
+ *   GICD @ 0x08000000 (distributor)
+ *   GICC @ 0x08010000 (CPU interface, MMIO - no ICC_* system registers)
+ * Same architecture as SUPERBIRD's GIC-400 - HAS_GIC_V3 must stay
+ * undefined here. The physical bases are injected into
+ * pexpert/gic/gic.zig at build-graph time (tools/zig/boards.zig
+ * GicBases), not read from this header; they are restated here only for
+ * documentation.
+ */
+#define GIC_SPURIOUS_IRQ          1023
+
+#define GICD_PHYS_BASE            0x08000000ULL
+#define GICD_SIZE                 0x1000
+#define GICC_PHYS_BASE            0x08010000ULL
+#define GICC_SIZE                 0x1000
+
+#define GICD_CTLR                 0x0
+#define GICD_CTLR_ENABLEGRP0      0x1
+#define GICD_CTLR_ENABLEGRP1      0x2
+
+#define GICC_CTLR                 0x0
+#define GICC_PMR                  0x4
+#define GICC_BPR                  0x8
+#define GICC_IAR                  0xc
+#define GICC_EOIR                 0x10
+#define GICC_CTLR_ENABLEGRP0      0x1
+#define GICC_CTLR_ENABLEGRP1      0x2
+
+#endif /* ! _PEXPERT_ARM64_QEMU_H */
diff --git a/pexpert/pexpert/arm64/SUPERBIRD.h b/pexpert/pexpert/arm64/SUPERBIRD.h
new file mode 100644
index 00000000..df31184f
--- /dev/null
+++ b/pexpert/pexpert/arm64/SUPERBIRD.h
@@ -0,0 +1,95 @@
+/*
+ * Board support header for the Spotify Car Thing ("superbird"), Amlogic
+ * S905D2 / G12A, quad Cortex-A53 (ARMv8.0-A, no PAC, no LSE, no FEAT_PAN3).
+ *
+ * Unlike the Apple SoC headers in this directory, this board has none of
+ * Apple's silicon: no AMCC read-only-region controller, no paravirtualized
+ * PAC/CTRR, no GICv3, no proprietary SMP bring-up registers. Everything
+ * gated behind those features must stay off.
+ */
+
+#ifndef _PEXPERT_ARM64_SUPERBIRD_H
+#define _PEXPERT_ARM64_SUPERBIRD_H
+
+#define NO_MONITOR                1
+#define SUPERBIRD                 1
+/*
+ * Generic flag for non-Apple-silicon boards: no Tightbeam/exclaves (no
+ * SEP/secure-enclave coprocessor to talk to), no per-SoC blob. Reused by
+ * any future open-hardware board, not just this one.
+ */
+#define OSS_HARDWARE               1
+#define NO_ECORE                  1
+
+/*
+ * No Apple Silicon PAC/CTRR/SME/PAN3 - Cortex-A53 is plain ARMv8.0-A.
+ * Deliberately do NOT define CPU_HAS_APPLE_PAC, HAS_PARAVIRTUALIZED_PAC,
+ * HAS_PARAVIRTUALIZED_CTRR, HAS_ARM_FEAT_SSBS2/SME/SME2/PAN3: none exist
+ * on this CPU and defining them would emit instructions that #UD here.
+ */
+
+/*
+ * Deliberately left UNDEFINED, not `#define __ARM_16K_PG__ 0` - see
+ * pexpert/pexpert/arm64/QEMU.h's comment: osfmk/arm64/proc_reg.h mixes
+ * `#ifdef __ARM_16K_PG__` and `#if __ARM_16K_PG__` guards, which
+ * disagree if this is defined-but-zero, producing a self-inconsistent
+ * TCR_EL1 (16K-style T0SZ with a 4K TG0) that instruction-aborts at
+ * SCTLR_EL1.M-enable time.
+ */
+#define __ARM_RANGE_TLBI__        0
+
+#define ARM_PARAMETERIZED_PMAP    1
+
+#include <pexpert/arm64/apple_arm64_common.h>
+#undef  BTI_ENFORCED
+#define BTI_ENFORCED 0
+#undef  __ARM64_PMAP_SUBPAGE_L1__
+#undef  __ARM64_PMAP_KERN_SUBPAGE_L1__
+/* Crypto extensions unconfirmed on this SoC's A53 SKU - use generic C. */
+#undef  __ARM_V8_CRYPTO_EXTENSIONS__
+/* This is not Apple silicon: no coherent-fabric assumptions, no Apple
+ * cache-maintenance helpers (nopreempt/mva_ops), no eng-fused sysctl. */
+#undef  APPLE_ARM64_ARCH_FAMILY
+
+/*
+ * Console: none. The Car Thing exposes no accessible UART; the only
+ * available output path at boot is the LCD framebuffer u-boot's bootxnu
+ * hands off via xnu_boot_arguments.video_information. Do not define
+ * PL011_UART / APPLE_UART / DOCKCHANNEL_UART - no serial driver exists
+ * for this board and none is wired up.
+ */
+#ifndef ASSEMBLER
+#define PLATFORM_PANIC_LOG_DISABLED
+#endif /* ! ASSEMBLER */
+
+/*
+ * GIC-400 (GICv2) interrupt controller, per meson-g12-common.dtsi:
+ *   GICD @ 0xffc01000 (distributor)
+ *   GICC @ 0xffc02000 (CPU interface, MMIO - no ICC_* system registers)
+ *   GICH @ 0xffc04000 (hypervisor interface, unused)
+ *   GICV @ 0xffc06000 (virtual CPU interface, unused)
+ * This is architecturally different from the GICv3 code in
+ * pexpert/arm/pe_fiq.c and osfmk/arm64/sleh.c (which assume ICC_* sysreg
+ * + redistributor access) - HAS_GIC_V3 MUST stay undefined here, and a
+ * GICv2 MMIO variant of that interrupt bring-up/EOI path is still needed.
+ */
+#define GIC_SPURIOUS_IRQ          1023
+
+#define GICD_PHYS_BASE            0xffc01000ULL
+#define GICD_SIZE                 0x1000
+#define GICC_PHYS_BASE            0xffc02000ULL
+#define GICC_SIZE                 0x2000
+
+#define GICD_CTLR                 0x0
+#define GICD_CTLR_ENABLEGRP0      0x1
+#define GICD_CTLR_ENABLEGRP1      0x2
+
+#define GICC_CTLR                 0x0
+#define GICC_PMR                  0x4
+#define GICC_BPR                  0x8
+#define GICC_IAR                  0xc
+#define GICC_EOIR                 0x10
+#define GICC_CTLR_ENABLEGRP0      0x1
+#define GICC_CTLR_ENABLEGRP1      0x2
+
+#endif /* ! _PEXPERT_ARM64_SUPERBIRD_H */
diff --git a/pexpert/pexpert/arm64/VMAPPLE.h b/pexpert/pexpert/arm64/VMAPPLE.h
index 08b35780..f6804c1c 100644
--- a/pexpert/pexpert/arm64/VMAPPLE.h
+++ b/pexpert/pexpert/arm64/VMAPPLE.h
@@ -53,6 +53,8 @@
 
 
 #include <pexpert/arm64/apple_arm64_common.h>
+#undef  BTI_ENFORCED
+#define BTI_ENFORCED 0
 #undef  __ARM64_PMAP_SUBPAGE_L1__
 #undef  __ARM64_PMAP_KERN_SUBPAGE_L1__
 
diff --git a/pexpert/pexpert/arm64/board_config.h b/pexpert/pexpert/arm64/board_config.h
index 5a6cd336..53833877 100644
--- a/pexpert/pexpert/arm64/board_config.h
+++ b/pexpert/pexpert/arm64/board_config.h
@@ -319,6 +319,54 @@
 #endif  /* ARM64_BOARD_CONFIG_VMAPPLE */
 
 
+#ifdef ARM64_BOARD_CONFIG_SUPERBIRD
+#include <pexpert/arm64/SUPERBIRD.h>
+
+#define MAX_L2_CLINE                   6  /* Cortex-A53: 64-byte L2 lines */
+#define MAX_CPUS                       1  /* v1: boot single-core; no PSCI SMP bring-up in xnu yet */
+#define MAX_CPU_CLUSTERS               1
+
+/*
+ * Do NOT define USE_APPLEARMSMP: that path drives Apple's proprietary
+ * CPU_OVRD/ACC power-manager registers, which don't exist on this SoC.
+ * Secondary cores are left parked (never issue PSCI CPU_ON).
+ */
+#define NO_CPU_OVRD                    1 /* CPU_OVRD register accesses are banned - no such register here */
+
+#define NO_XNU_PLATFORM_ERROR_HANDLER  1
+
+#endif  /* ARM64_BOARD_CONFIG_SUPERBIRD */
+#ifdef ARM64_BOARD_CONFIG_IPAD41
+#include <pexpert/arm64/IPAD41.h>
+
+#define MAX_L2_CLINE                   6  /* Cyclone: 64-byte L2 lines */
+#define MAX_CPUS                       1  /* v1: boot CPU only; A7 dual-core bring-up TBD */
+#define MAX_CPU_CLUSTERS               1
+
+/* Same rationale as SUPERBIRD: no Apple CPU_OVRD/ACC power-manager
+ * registers are driven by this port. */
+#define NO_CPU_OVRD                    1
+
+#define NO_XNU_PLATFORM_ERROR_HANDLER  1
+
+#endif  /* ARM64_BOARD_CONFIG_IPAD41 */
+
+#ifdef ARM64_BOARD_CONFIG_QEMU
+#include <pexpert/arm64/QEMU.h>
+
+#define MAX_L2_CLINE                   6  /* Cortex-A53: 64-byte L2 lines */
+#define MAX_CPUS                       1  /* v1: boot single-core; no PSCI SMP bring-up in xnu yet */
+#define MAX_CPU_CLUSTERS               1
+
+/* Same rationale as SUPERBIRD: no Apple CPU_OVRD/ACC power-manager
+ * registers exist on QEMU's `virt` machine either. */
+#define NO_CPU_OVRD                    1
+
+#define NO_XNU_PLATFORM_ERROR_HANDLER  1
+
+#endif  /* ARM64_BOARD_CONFIG_QEMU */
+
+
 
 #ifndef HAS_UNCORE_CTRS
 #undef UNCORE_VERSION
diff --git a/pexpert/pexpert/arm64/gic.h b/pexpert/pexpert/arm64/gic.h
new file mode 100644
index 00000000..b426104d
--- /dev/null
+++ b/pexpert/pexpert/arm64/gic.h
@@ -0,0 +1,30 @@
+/*
+ * Minimal GIC-400 (GICv2) MMIO driver interface for open-hardware ARM64
+ * boards (e.g. SUPERBIRD/Amlogic S905D2) that have no Apple Interrupt
+ * Controller and no GICv3 redistributor/ICC_* system registers. Register
+ * offsets/constants live in the board header (e.g. arm64/SUPERBIRD.h)
+ * since they are architecturally fixed by the GICv2 spec, not board data;
+ * only the physical base addresses are board-specific.
+ */
+#ifndef _PEXPERT_ARM64_GIC_H
+#define _PEXPERT_ARM64_GIC_H
+
+#ifndef ASSEMBLER
+
+#include <stdint.h>
+
+/* Called once, from PE_init_platform() after the VM is up (needs ml_io_map
+ * for the GICD/GICC MMIO windows). Maps the distributor + CPU interface,
+ * enables both interrupt groups, unmasks all priorities, and installs
+ * itself as the primary IRQ dispatcher via ml_install_interrupt_handler().
+ * No-op if GICD_PHYS_BASE is not defined for this board. */
+void gic_init(void);
+
+/* Enable (unmask) forwarding of the given GIC interrupt ID (SPI or PPI,
+ * i.e. the GIC-numbered IRQ, not the DT PPI/SPI-relative number) at the
+ * distributor. SGIs (0-15) are always enabled and need no call here. */
+void gic_enable_interrupt(uint32_t irq);
+
+#endif /* !ASSEMBLER */
+
+#endif /* !_PEXPERT_ARM64_GIC_H */
